src/compiler/types.ts

```diff
@@ -603,15 +603,75 @@ namespace ts {
 
     export type DeclarationName = Identifier | StringLiteral | NumericLiteral | ComputedPropertyName | BindingPattern;
 
-    export interface RealDeclaration extends Node {
+    export interface DeclarationBase extends Node {
         _declarationBrand: any;
         name?: DeclarationName;
     }
 
-    // Binary expressions can be declarations if they are 'exports.foo = bar' expressions in JS files
-    export type Declaration = RealDeclaration | BinaryExpression;
-
-    export interface DeclarationStatement extends RealDeclaration, Statement {
+    export type Declaration =
+        | ArrowFunction
+        // Binary expressions can be declarations if they are 'exports.foo = bar' expressions in JS files
+        | BinaryExpression

```

Comment:
This approach would require separate types for `this.x = 12` and `Foo.x = 12` and `Foo.prototype.x` assignments, wouldn't it? 

For the last two, there's no way to change the inherited type of `text`, since it can vary, though I think `JsPropertyPropertyAccess` could just have `expression: Identifier`.